tests/unit/core/mender-artifact_test.cpp

```diff
@@ -0,0 +1,320 @@
+}
+
+TEST_F(MenderArtifactTest, ProcessData) {
+    mender_artifact_ctx_t *ctx = mender_artifact_create_ctx(1024);
+
+    /* Process artifact data */
+    EXPECT_EQ(MENDER_OK, mender_artifact_process_data(ctx, data, artifact_data.size(), mock_download_data));
+
+    /* Number of payloads in the artifact, expect 1 */
+    EXPECT_EQ(ctx->payloads.size, 1);
+
+    /* Type of payload, expect `unit-test` */
+    mender_artifact_payload_t *payload = ctx->payloads.values;
+    EXPECT_STREQ(payload->type, "unit-test");
+
+    /* The original context size was too small, so the new one should be larger
+     * than the original isze */
+    EXPECT_GT(ctx->input.size, ctx->input.orig_size);
+
+    /* Should be 0/NULL, as there are no artifacts currently
+     * being parsed */
+    EXPECT_EQ(ctx->file.size, 0);
+    EXPECT_EQ(ctx->file.index, 0);
+    EXPECT_STREQ(ctx->file.name, NULL);
+
+    /* One thing should be provided, the artifact name */
+    mender_key_value_list_t *provides = ctx->artifact_info.provides;
+    EXPECT_STREQ(provides->value, "unit-test-artifact");
+
+    /* Should depend on one thing, the artifact type */
+    mender_key_value_list_t *depends = ctx->artifact_info.depends;
+    EXPECT_STREQ(depends->value, "unit-test");
+
+    /* Check the stream state */
+    EXPECT_EQ(ctx->stream_state, MENDER_ARTIFACT_STREAM_STATE_PARSING_HEADER);
+
+    mender_artifact_release_ctx(ctx);
+}
+
+TEST_F(MenderArtifactTest, ProcessData_TooLargeSize) {
+    mender_artifact_ctx_t *ctx = mender_artifact_create_ctx(1024);
+
+    /* If ctx->file.size (expected_size) is too big, the realloc will fail,
+     * and we'll attempt to realloc ctx->input.length + input_length (new_size)
+     * instead. This checks that the data is correctly copied to the internal buffer,
+     * even though the expected size is too big */
+    ctx->file.size = 99999999999999999;
+    EXPECT_EQ(MENDER_OK, mender_artifact_process_data(ctx, data, artifact_data.size(), mock_download_data));
+
+    mender_artifact_release_ctx(ctx);
+
+    ctx = mender_artifact_create_ctx(1024);
+
+    /* If the input size is too big, we're not able to realloc, and should fail */
+    ctx->input.length = 99999999999999999;
+    EXPECT_EQ(MENDER_FAIL, mender_artifact_process_data(ctx, data, artifact_data.size(), mock_download_data));
+
+    mender_artifact_release_ctx(ctx);
+}
+
+TEST_F(MenderArtifactTest, ProcessData_EmptyInput) {
+    mender_artifact_ctx_t *ctx = mender_artifact_create_ctx(1024);
+
+    EXPECT_EQ(MENDER_OK, mender_artifact_process_data(ctx, NULL, 0, mock_download_data));
+
+    EXPECT_EQ(ctx->input.size, ctx->input.orig_size);
+    EXPECT_EQ(ctx->file.size, 0);
+    EXPECT_EQ(ctx->file.index, 0);
+    EXPECT_STREQ(ctx->file.name, NULL);
+
+    mender_artifact_release_ctx(ctx);
+}
+
+TEST_F(MenderArtifactTest, ProcessData_InvalidData) {
+    mender_artifact_ctx_t *ctx = mender_artifact_create_ctx(1024);
+    string                 invalid_data(1240, 'A');
```

Comment:
I'd prefer `vector<uint8_t> invalid_data(1240, 'A')` and then using `invalid_data.data()`.